[WPE] WPE Platform: Dark mode setting does not work

carlosgcampos opened the pull request:
#### 696227e4897cf2cb14cc5d2ac3ae4f3bcab55ff8
<pre>
[WPE] WPE Platform: Dark mode setting does not work
<a href="https://bugs.webkit.org/show_bug.cgi?id=293066">https://bugs.webkit.org/show_bug.cgi?id=293066</a>

Reviewed by Adrian Perez de Castro.

Implement PageClientImpl::effectiveAppearanceIsDark() checking the value
of dark mode setting and notify the page that appearance has changed
when the setting changes.

Canonical link: <a href="https://commits.webkit.org/295000@main">https://commits.webkit.org/295000@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/907b9a1cde55d2fadc13c09a56e98ccb3e89e5c2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103782 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23484 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13804 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108975 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54434 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23834 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32030 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78847 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106788 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18497 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93623 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59181 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/18306 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53810 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/88077 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11729 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111362 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30938 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22809 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87849 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31299 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89824 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87503 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32390 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10112 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25295 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16843 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30866 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36169 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30660 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33995 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32221 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

spenap commented on Source/WebKit/UIProcess/API/wpe/WPEWebViewPlatform.cpp:
```diff
@@ -134,10 +135,16 @@ ViewPlatform::ViewPlatform(WPEDisplay* display, const API::PageConfiguration& co
     m_pageProxy->initializeWebPage(pageConfiguration.openedSite(), pageConfiguration.initialSandboxFlags());
+
+    WebCore::SystemSettings::singleton().addObserver([this](const auto& state) {
+        if (state.darkMode)
```
I understand the logic in `effectiveAppearanceIsDark()`, returning `value_of(false)`. 

But here, what happens if we were in dark mode and we turn to light mode? this check would fail so we wouldn't call `didChange`. Is the transition from dark to light handled elsewhere?

aperezdc replied:
The `state` only tells which settings have changed, not their actual values. When this calls `page().effectiveAppearanceDidChange()` then WebKit will call again the `PageClientImpl::effectiveAppearanceIsDark()` function to re-read the value.

spenap replied:
Ah! This `darkMode` is an optional! Then yes, the test is true when a value (true or false) is set, I see :-)

carlosgcampos committed:
[WPE] WPE Platform: Dark mode setting does not work
https://bugs.webkit.org/show_bug.cgi?id=293066

Reviewed by Adrian Perez de Castro.

Implement PageClientImpl::effectiveAppearanceIsDark() checking the value
of dark mode setting and notify the page that appearance has changed
when the setting changes.

Canonical link: https://commits.webkit.org/295000@main
